src/treelearner/cuda_tree_learner.cpp

```diff
@@ -0,0 +1,1049 @@
+/*!
+ * Copyright (c) 2020 IBM Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ */
+#ifdef USE_CUDA
+#include "cuda_tree_learner.h"
+
+#include <LightGBM/utils/array_args.h>
+#include <LightGBM/network.h>
+#include <LightGBM/bin.h>
+
+#include <pthread.h>
+
+#include <LightGBM/cuda/cuda_utils.h>
+
+#include <algorithm>
+#include <vector>
+
+#include "../io/dense_bin.hpp"
+
+#define cudaMemcpy_DEBUG 0  // 1: DEBUG cudaMemcpy
+#define ResetTrainingData_DEBUG 0  // 1: Debug ResetTrainingData
+
+#define GPU_DEBUG 0
+
+static void *launch_cuda_histogram(void *thread_data) {
+  ThreadData td = *(reinterpret_cast<ThreadData*>(thread_data));
+  int device_id = td.device_id;
+  CUDASUCCESS_OR_FATAL(cudaSetDevice(device_id));
+
+  // launch cuda kernel
+  cuda_histogram(td.histogram_size,
+                td.leaf_num_data, td.num_data, td.use_all_features,
+                td.is_constant_hessian, td.num_workgroups, td.stream,
+                td.device_features,
+                td.device_feature_masks,
+                td.num_data,
+                reinterpret_cast<uint*>(td.device_data_indices),
+                td.leaf_num_data,
+                td.device_gradients,
+                td.device_hessians, td.hessians_const,
+                td.device_subhistograms, td.sync_counters,
+                td.device_histogram_outputs,
+                td.exp_workgroups_per_feature);
+
+  CUDASUCCESS_OR_FATAL(cudaGetLastError());
+
+  return NULL;
+}
+
+/*
+static void *wait_event(void *wait_obj) {
+  CUDASUCCESS_OR_FATAL(cudaEventSynchronize(*(cudaEvent_t *)wait_obj));
+}*/
+
+namespace LightGBM {
+
+CUDATreeLearner::CUDATreeLearner(const Config* config)
+  :SerialTreeLearner(config) {
+  use_bagging_ = false;
+  nthreads_ = 0;
+  if (config->gpu_use_dp && USE_DP_FLOAT) {
```

Comment:
It seems so far we have `USE_DP_FLOAT` always be 1. 
So is the SP float part actually tested? I understand that so far the CUDA code does not support changing precision at runtime. But if the SP float indeed works, can we change `USE_DP_FLOAT` at compile time and produce a single precision CUDA build?

Reply:
The SP version of CUDA has not been implemented as of yet.  Most of this code is in anticipation of future work.